Add QV plotter for plotting heavy output probability vs ntrials

HuangJunye opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Add QV plotter for plotting heavy output probability vs ntrials similar to Figure 2a of [QV64 paper](https://arxiv.org/abs/2008.08571).

### Details and comments
Fixes: #213 and #483:
1. add `plot_hop_accumulative` method for plotting heavy output probability vs ntrials
1. plots individual HOP and accumulative HOP
1. plots 2-sigma shaded area
1. plots 2/3 success threshold

Example:
```qv_fitter.plot_hop_accumulative(depth=5)``` with 100 trials
![QV32](https://user-images.githubusercontent.com/7631333/92329839-d3f78580-f09c-11ea-80aa-c5002cc8c52b.png)
```qv_fitter.plot_hop_accumulative(depth=6)``` with 1000 trials
![qv64](https://user-images.githubusercontent.com/7631333/92330301-677e8580-f0a0-11ea-807c-cb309e4eaaa0.png)

HuangJunye commented:
The lint test error is due to `label='2$\sigma$'`. This is only a warning why does it fail the test?
```
qiskit/ignis/verification/quantum_volume/fitters.py:355:68: W605 invalid escape sequence '\s'
ERROR: InvocationError for command /home/runner/work/qiskit-ignis/qiskit-ignis/.tox/lint/bin/pycodestyle qiskit/ignis test/ (exited with code 1)
```

HuangJunye commented:
Suggest reviewer: @dcmckayibm

HuangJunye commented:
@ShellyGarion Can you also shed some lights on how to fix the tests errors in this PR? The python tests, doc test and tutorials test give this error message `No module named 'qiskit.circuit.library.template_circuits'` which seems to be a Terra problem. I am not sure why this is happening. 

lint error is because '\s' in '\sigma' gives a warning but it shouldn't fail the test. Is there anyway to work around to allow '\s'?

mtreinish commented:
> @ShellyGarion Can you also shed some lights on how to fix the tests errors in this PR? The python tests, doc test and tutorials test give this error message `No module named 'qiskit.circuit.library.template_circuits'` which seems to be a Terra problem. I am not sure why this is happening.
> 
> lint error is because '\s' in '\sigma' gives a warning but it shouldn't fail the test. Is there anyway to work around to allow '\s'?

For the lint error you either need to escape the `\` as `\\s` or change the string to be a raw string with `r"\s"` to tell python to not treat `\s` as an escape sequence. Either approach is valid, but without that in certain contexts python will try to use that as an escape sequence which is invalid and why the linter fails.

mtreinish reviewed (commented):
Also while fixing this make sure to add a release note for this new plotter function. See: https://github.com/Qiskit/qiskit-ignis/blob/master/CONTRIBUTING.md#release-notes for more details.

HuangJunye commented:
> @ShellyGarion Can you also shed some lights on how to fix the tests errors in this PR? The python tests, doc test and tutorials test give this error message `No module named 'qiskit.circuit.library.template_circuits'` which seems to be a Terra problem. I am not sure why this is happening.
> 
> lint error is because '\s' in '\sigma' gives a warning but it shouldn't fail the test. Is there anyway to work around to allow '\s'?

I changed '\\s'. The lint error is now fixed. All tests are passed. Thank you! Release note is also added. Can you please review?

ShellyGarion commented on releasenotes/notes/plot-hop-vs-ntrials-aa4b8d80808edfdb.yaml:
```diff
@@ -0,0 +1,8 @@
+---
+features:
+  - |
+    Add :meth:`qiskit.ignis.verification.quantum_volume.QVFitter.plot_hop_accumulative` for plotting heavy output probability (HOP) vs number of trails similar to Figure 2a of QV64 paper (arXiv:2008.08571).
```
Looks good - only a misprint:
trails --> trials

HuangJunye replied:
Thank you :) I fixed the typos. Otherwise the code is good?

ShellyGarion reviewed (changes requested):
Looks good - only a misprint in release notes.

HuangJunye commented:
Thanks for the approval! @dcmckayibm and @ShellyGarion

mtreinish commented on qiskit/ignis/verification/quantum_volume/fitters.py:
```diff
@@ -308,6 +313,72 @@ def plot_qv_data(self, ax=None, show_plt=True):
+        labels = [labels[1], labels[2], labels[0], labels[3]]
+        ax.legend(handles, labels)
+
+        plt.close(fig)  # close additional figure
```
This is the same comment as on #495 https://github.com/Qiskit/qiskit-ignis/pull/495/files#r492960737 we shouldn't be unconditionally closing here.

HuangJunye replied:
I have made the suggested changes.

mtreinish commented on releasenotes/notes/plot-hop-vs-ntrials-aa4b8d80808edfdb.yaml:
```diff
@@ -0,0 +1,8 @@
+---
+features:
+  - |
+    Add :meth:`qiskit.ignis.verification.quantum_volume.QVFitter.plot_hop_accumulative` for plotting heavy output probability (HOP) vs number of trials similar to Figure 2a of QV64 paper (arXiv:2008.08571).
+    * HOP of individual trials are plotted as scatters.
+    * Cummulative HOP are plotted in red line.
+    * Two-sigma confidence intervals are plotted as shaded area.
+    * 2/3 success threshold is plotted as dashed line.
```
Can you reformat this slightly. This will look quite weird in the rendered release notes:

![Screenshot_2020-09-22_14-54-41](https://user-images.githubusercontent.com/2447371/93925069-91ac8600-fce3-11ea-9e25-c16b1ac4c219.png) (which I just pulled from the artifact link on the ci results from the docs run: https://github.com/Qiskit/qiskit-ignis/pull/493/checks?check_run_id=1127699605)

Some quick suggestions, add a link for the arXiv paper and rework the bulleted list as sentences explaining how the factor into the visualization. If you want a bulleted list you'll need to add a newline before the first line (https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#lists-and-quote-like-blocks ). Also can you wrap the lines at 100 characters, having a super long line like this will cause issues when this gets integrated into the combined documentation.

HuangJunye replied:
I made the suggested changes. Please let me know if the release note is still not good.

mtreinish reviewed (approved):
LGTM, thanks for makign the updates
